Chap_Terms.tex

```diff
@@ -4,39 +4,30 @@
+The following terms and terminology is used throughout this document:
 \begin{itemize}
-\item \declareterm{session}\emph{session} refers to an allocated set of resources assigned to a particular user by the system \ac{WLM}. Historically, \ac{HPC} sessions have consisted of a static allocation of resources - i.e., a block of resources are assigned to a user in response to a specific request and managed as a unified collection. However, this is changing in response to the growing use of dynamic programming models that require on-the-fly allocation and release of system resources. Accordingly, the term \emph{session} in this document refers to the current block of assigned resources and is a potentially dynamic entity.
-\item \declareterm{slot}\declareterm{slots}\emph{slot} refers to an allocated entry for a process. \acp{WLM} frequently allocate entire nodes to a \emph{session}, but can also be configured to define the maximum number of processes that can simultaneously be executed on each node. This often corresponds to the number of hardware \acp{PU} (typically cores, but can also be defined as hardware threads) on the node. However, the correlation between hardware \acp{PU} and slot allocations strictly depends upon system configuration.
-\item \declareterm{job}\emph{job} refers to a set of one or more \emph{applications} executed as a single invocation by the user within a session. For example, ``\textit{mpiexec -n 1 app1 : -n 2 app2}'' is considered a single \ac{MPMD} job containing two applications.
+\item \declareterm{session}\emph{session} refers to an allocated set of resources assigned to a particular user by the system \ac{WLM}. Historically, \ac{HPC} \emph{sessions} have consisted of a static allocation of resources - i.e., a block of resources are assigned to a user in response to a specific request and managed as a unified collection. However, this is changing in response to the growing use of dynamic programming models that require on-the-fly allocation and release of system resources. Accordingly, the term \emph{session} in this document refers to the current block of assigned resources and is a potentially dynamic entity.
+\item \declareterm{slot}\declareterm{slots}\emph{slot} refers to an allocated set of resources assigned to a process. \acp{WLM} frequently allocate entire nodes to a \emph{session}, but can also be configured to define the maximum number of processes that can simultaneously be executed on each node. This often corresponds to the number of hardware \acp{PU} (typically cores, but can also be defined as hardware threads) on the node. However, the correlation between hardware \acp{PU} and slot allocations strictly depends upon system configuration.
+\item \declareterm{application}\emph{application} refers to a single executable (binary, script, etc.) member of a \emph{job}. Applications consist of one or more \emph{processes}, either operating independently or cooperatively at any given time during their execution.
```

Comment:
Processes -> do you mean OS processes here? Seems implementation specific and too narrowing to me. What if someone starts designing a system that works only on thread basis (e.g., how it has happened with MPI and MPC) - I don 't think we want to prevent that

Reply:
@schulzm I agree that this section would benefit from a definition of what we mean by "process". I also agree (I think this is in a different comment somewhere below) that keeping the definition abstract (and not tied to an OS process) will keep the door open for different kinds of implementations in the future (e.g. threaded)

Reply:
We use the term `process` extensively in the standard, so if we want to broaden its definition to mean "execution unit" to allow for a thread interpretation then we would definitely need to define it here. However, that would be a new perspective on the interface and might be much broader reaching that just Chapter 2.

My suggestion would be to take the (maybe implied) definition of a unix process in this PR, then follow up with another that expands that definition so that it can be debated and refined on its own.

Reply:
From the discussion in the ASC meeting, I agree with @jjhursey that broadening the scope beyond OS processes would be a much more extensive change to the document (and thus a separate issue if needed). I am not sure if the meaning of OS processes is 100% clear from this writing. However, I suppose without explicitly saying that 'processes' means something different than OS processes, one should assume OS processes. If we did clarify it, I am not even sure what terminology would be most appropriate.

Reply:
Is it really an OS process, though? Really intended as a question, but as an application is a set of processes, and a job is a set of applications, what does this mean for heterogeneous systems? How do you represent parts of the applications run on accelerators (both within nodes, but also run on self hosted boosters)? Also, what does this mean for MPC applications? Does for me, as a user, the terminology change when I want to run a 20 MPI process MPC code vs. a 20 MPI process MPICH code?

Reply:
Created an issue to track this: #262

Reply:
(my) Notes from Aug. 24, 2020 meeting on this thread:
 * v4 draft document added the following items:
 > **process** refers to an operating system process, also commonly referred to as a heavyweight process. A process is often comprised of multiple lightweight threads, commonly known as simply threads.
 > 
 > **node** refers to a single operating system instance. Note that this may encompass one or more physical objects.
 * For a BlueGene system, there is no process separate from the OS. We should make sure the definition encompasses that type of environment.
 * If you `PMIx_Spawn` a program. That program then calls fork(), and both the parent and the child process both call `PMIx_Init`. Is this a supported scenario? (Current implementation would not allow this - the question is should we?) If so, then from the server's perspective it created 1 process, but now two processes are connecting back - should it give them two different ranks? This gets tricky... Maybe something to think about for the future.

Reply:
Just clarifying: the current implementation does indeed support the case where a parent forks a child and both call `PMIx_Init`.

Note that the PMIx server does not assign ranks - that is left to the host environment. The PMIx server is required to detect that multiple clients have connected using the same rank and track them. OpenPMIx does this today, correctly declaring that the "client has disconnected" once all "clones" of that rank have done so. I can update the client and server chapters to make that clear.